### PR TITLE
feature: add block pair token price entity

### PR DESCRIPTION
**Pull request by berteotti**
This feature adds BlockPairTokenPrice entity so we can fetch and sort Pair Token prices by block timestamp and number.

### CODE REVIEW

File: src/mappings/core.ts [hunk 1]
```diff
@@ -586,3 +594,24 @@ export function handleSwap(event: Swap): void {
...
+
+export function handleBlockPairTokenPrice(block: ethereum.Block): void {
+  let pairContractAddress = dataSource.address()
+  let pair = Pair.load(pairContractAddress.toHexString())
+
+  if (!pair) {
+    return
+  }
+
+  let blockPairTokenPrice = new BlockPairTokenPrice(block.hash.toHexString())
```
Comment (adamazad) [inline src/mappings/core.ts:606]: For each block, the total number of `BlockPairTokenPrice` entities created equals the number of Swapr pairs the protocol has at that block.

In this implementation, all BlockPairTokenPrice share the same ID which is an issue. A simple uniqueness solution is using block hash or block height with the pair address:


```suggestion
  let blockPairTokenPrice = new BlockPairTokenPrice(block.number.toString() + '-' + pairContractAddress.toHexString())
```

This produces IDs like: `15061515-0x1865d5445010e0baf8be2eb410d3eae4a68683c2`, `15061516-0x1865d5445010e0baf8be2eb410d3eae4a68683c2`, `15061517-0x1865d5445010e0baf8be2eb410d3eae4a68683c2` and so on

File: schema.graphql [hunk 2]
```diff
@@ -475,3 +475,14 @@ type SingleSidedStakingCampaignPosition @entity {
...
+type BlockPairTokenPrice @entity {
+  id: ID!
+  blockNumber: BigInt!
+  blockTimestamp: BigInt!
+  pair: Pair!
+  token0Price: BigDecimal!
+  token0Address: Bytes!
+  token1Price: BigDecimal!
+  token1Address: Bytes!
+}
```
Comment (adamazad) [inline schema.graphql:488]: Missing newline at end of file. :)

**Review by adamazad — CHANGES_REQUESTED** [review]

File: src/mappings/core.ts (hunk 1 above)
Comment (berteotti) [reply src/mappings/core.ts]: Thanks Adam, it's now fixed in [fe8aeca](https://github.com/dxgraphs/swapr-subgraph/pull/23/commits/fe8aecae915fe290462a0f652681a4a6f51b3575)

File: schema.graphql (hunk 2 above)
Comment (berteotti) [reply schema.graphql]: Fixed in [fe8aeca](https://github.com/dxgraphs/swapr-subgraph/pull/23/commits/fe8aecae915fe290462a0f652681a4a6f51b3575)

File: schema.graphql [hunk 3]
```diff
@@ -475,3 +474,58 @@ type SingleSidedStakingCampaignPosition @entity {
...
+type FiveMinutesPairTokenPrice implements PairTokenPrice @entity {
+  id: ID!
+  blockNumber: BigInt!
+  blockTimestamp: BigInt!
+  pair: Pair!
+  token0Price: BigDecimal!
+  token0Address: Bytes!
+  token1Price: BigDecimal!
+  token1Address: Bytes!
+}
```
Comment (adamazad) [inline schema.graphql:498]: I went back to an old subgraph I created. [There's support for enums in GraphQL](https://github.com/dxgraphs/aqua-subgraph/blob/da403972c9c26dcbcea3c0ff04e09cb9c3d522ba/schema.graphql#L41-L58):

```typescript
enum PairTokenPriceTimeFrame {
  FIVE_MINUTES,
  FIFTEEN_MINUTES,
  ONE_HOUR,
  TWELVE_HOURS
}

type PairTokenPrice  @entity {
  id: ID!
  timeframe: PairTokenPriceTimeFrame!
  blockNumber: BigInt!
  blockTimestamp: BigInt!
  pair: Pair!
  token0Price: BigDecimal!
  token0Address: Bytes!
  token1Price: BigDecimal!
  token1Address: Bytes!
}
```

The rest of entities are unnecessary if we implement enums.

File: src/mappings/helpers.ts [hunk 4]
```diff
@@ -346,3 +346,13 @@ export function addDailyUniqueAddressInteraction(event: ethereum.Event, address:
...
+
+export function createPairTokenPrice(pairTokenPrice: Entity, block: ethereum.Block, pair: Pair): void {
+  pairTokenPrice.set('blockNumber', Value.fromBigInt(block.number))
+  pairTokenPrice.set('blockTimestamp', Value.fromBigInt(block.timestamp))
+  pairTokenPrice.set('pair', Value.fromString(pair.id))
+  pairTokenPrice.set('token0Price', Value.fromBigDecimal(pair.token0Price || ZERO_BD))
+  pairTokenPrice.set('token1Price', Value.fromBigDecimal(pair.token1Price || ZERO_BD))
+  pairTokenPrice.set('token0Address', Value.fromAddress(Address.fromString(pair.token0 || ADDRESS_ZERO)))
+  pairTokenPrice.set('token1Address', Value.fromAddress(Address.fromString(pair.token1 || ADDRESS_ZERO)))
+}
```
Comment (adamazad) [inline src/mappings/helpers.ts:358]: Looks good :)

**Review by adamazad — CHANGES_REQUESTED** [review]

File: schema.graphql (hunk 3 above)
Comment (berteotti) [reply schema.graphql]: Thanks for the tip, just updated the entity in this commit [a8a6fc4](https://github.com/dxgraphs/swapr-subgraph/pull/23/commits/a8a6fc4956b216c3ed0b3d49652dca5724e8d158)

File: src/mappings/helpers.ts [hunk 5]
```diff
@@ -346,3 +347,25 @@ export function addDailyUniqueAddressInteraction(event: ethereum.Event, address:
     dailyUniqueAddressInteractionsData.save()
   }
 }
+
+export abstract class PairTokenPriceTimeframe {
+  static FIVE_MINUTES: string = 'FIVE_MINUTES'
+  static FIFTEEN_MINUTES: string = 'FIFTEEN_MINUTES'
+  static ONE_HOUR: string = 'ONE_HOUR'
+  static TWELVE_HOURS: string = 'TWELVE_HOURS'
+}
```
Comment (adamazad) [inline src/mappings/helpers.ts:356]: Great substitute for enums in AsseyblyScript :)

**Review by adamazad — APPROVED** [review]

**Pull request merged**